cmd/mysqlout: add utility for converting mysql OUTFILE to CSV

dt opened the pull request:
Out IMPORT functionality only works on spec-compliant CSVs.

MySQL's OUTFILE (as used in mysqldump --tab) can be configured to almost
CSV, but its escaping and field enclosing behavior does not quite match
the CSV spec. Specifically, CSV escaping is via an extra quote in front
of quote literals, with separators (line or field) appearing inside of
quotes being treated as literal without escaping. MySQL escapes either
always or not at all, rather than considering field enclosures. This
makes it harder to use its options to make something that parses as CSV
if the data actually includes quotes and separators.

This adds a small program that implements the mySQL decoding logic to
produce a CSV that is compatible with IMPORT. Eventually the plan is
to bring this logic into the IMPORT command itself as an alternative
format to CSV. For now though it is offered as a prototype in a
standalone cli utility that converts (and chunks) to CSV.

Release note: none.

dt committed:
cmd/mysqlout: add utility for converting mysql OUTFILE to CSV

Out IMPORT functionality only works on spec-compliant CSVs.

MySQL's OUTFILE (as used in mysqldump --tab) can be configured to almost
CSV, but its escaping and field enclosing behavior does not quite match
the CSV spec. Specifically, CSV escaping is via an extra quote in front
of quote literals, with separators (line or field) appearing inside of
quotes being treated as literal without escaping. MySQL escapes either
always or not at all, rather than considering field enclosures. This
makes it harder to use its options to make something that parses as CSV
if the data actually includes quotes and separators.

This adds a small program that implements the mySQL decoding logic to
produce a CSV that is compatible with IMPORT. Eventually the plan is
to bring this logic into the IMPORT command itself as an alternative
format to CSV. For now though it is offered as a prototype in a
standalone cli utility that converts (and chunks) to CSV.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25287)
<!-- Reviewable:end -->

mjibson commented on pkg/cmd/mysqlout-to-csv/main.go:
```diff
@@ -0,0 +1,130 @@
+	"github.com/cockroachdb/cockroach/pkg/util"
+)
+
+var lineSep = flag.String("lines-terminated-by", "\n", "line separator character")
```
maybe make a `var (` block here

dt replied:
Done.

mjibson commented on pkg/cmd/mysqlout-to-csv/main.go:
```diff
@@ -0,0 +1,130 @@
+
+var lineSep = flag.String("lines-terminated-by", "\n", "line separator character")
+var fieldsSep = flag.String("fields-terminated-by", "\t", "field separator character")
+var encloseChar = flag.String("fields-enclosed-by", "", "field enclosing character")
```
does this also need an "is optional" flag to support `FIELDS [OPTIONALLY] ENCLOSED BY`? or maybe this parser can tell if it was enclosed or not with a lookahead or something? need to test this.

mjibson commented on pkg/cmd/mysqlout-to-csv/main.go:
```diff
@@ -0,0 +1,130 @@
+
+var progressFreq = flag.Int("progress", 1000, "number of rows per progress update (0 for no updates)")
+
+var bufSizeKb = flag.Int("buf-size-kb", 1024, "buffer size")
```
does this really need to be an option?

dt replied:
I guess not, I was experimenting with it to see if 1kb vs 1mb vs 100mb made a meaningful difference in time to convert our 180mb test case. It didn't on a MBP SSD, so I'll switch it to 64kb.

mjibson commented on pkg/cmd/mysqlout-to-csv/main.go:
```diff
@@ -0,0 +1,130 @@
+	writer.Flush()
+	out.Close()
+
+	fmt.Printf("\rWrote %d rows.\n", totalRows)
```
Do we care about supporting output to stdout? Even if we don't right now it might make sense to use log instead of fmt to not cross the streams.

mjibson commented on pkg/cmd/mysqlout-to-csv/main.go:
```diff
@@ -0,0 +1,130 @@
+		os.Exit(1)
+	}
+
+	in, err := os.Open(flag.Arg(0))
```
should we support stdin?

mjibson commented on pkg/cmd/mysqlout-to-csv/convert.go:
```diff
@@ -0,0 +1,160 @@
+			c, w := utf8.DecodeRune(buf[i:])
+			if c == utf8.RuneError && w == 1 {
+				remaining := total - i
+				//				fmt.Printf("\n\n i %d n %d r total %d remaining %d\n\n", i, n, total, remaining)
```
leftover debug?

dt replied:
Done.

mjibson commented on pkg/cmd/mysqlout-to-csv/convert.go:
```diff
@@ -0,0 +1,160 @@
+		start = 0
+
+		for i := 0; i < total; {
+			c, w := utf8.DecodeRune(buf[i:])
```
Is it safer to use bufio.Scanner with bufio.ScanRunes? I *think* this code here can error if the Read boundary happens to be in the middle of a valid multi-byte rune.

dt replied:
Added several comments describing that process.

madelynnblue reviewed (commented):
I want to see a test suite with some adversarial inputs and various mysql outfile options on them corresponding to some CSVs. I would like to see NULLs, weirdo but valid unicode strings, rando BYTES data, things with \n, \r, `"`, `,`, etc.

dt reviewed (commented):
Yeah, I still need to write some tests, just wanted early feedback.

dt commented:
@mjibson so while `bufio.Scanner` and `ScanRunes` were too slow, `bufio.Reader` and `ReadRune` is basically exactly what I have here: `bufio.Reader.fill()` does exactly the same "slide the remaining data to beginning of buf via a copy() and then call Read() on the offset slice to fill the remainder`  thing I was doing, and the `ReadRune` method calls fill if there aren't enough bytes left in the buffer to attempt decode rune (plus it has a fast-path for simple runes). Performance appears to be exactly the same using that as my hand-rolled buffered reader, so i've switched to it.

dt committed:
cmd/mysqlout: add conversion tests

This tests our ability to round-trip strings from go, into mysql and then back via a mysqldump.
Release note: none.

dt commented:
@mjibson Added a few tests that use `mysqldump`'ed testdata. RFAL.

madelynnblue reviewed (commented):
This is looking great.

dt commented on pkg/cmd/mysqlout-to-csv/convert.go:
```diff
@@ -0,0 +1,156 @@
+				if err := d.row(row); err != nil {
+					return err
+				}
+				row = make([]string, 0, len(row))
```
yeah, or even as-is, it isn't clear that `RowFunc` made its own copy

dt commented on pkg/cmd/mysqlout-to-csv/convert_test.go:
```diff
@@ -0,0 +1,197 @@
+			name: "escape-quote-csv-no-enclose",
+			cfg:  dumpReader{hasEscapeChar: true, escapeChar: '"', fieldSep: ',', rowSep: '\n'},
+			null: `\N`,
+		},
```
mysql has no escape char by default so AFAIK, there is no way that works if your data contains your special chars.

dt committed:
cmd/mysqlout: passthrough non-unicode bytes

Non-unicode bytes would previously produce an error.

However, since our delimiters are defined as valid unicode runes, if a
byte fails to decode as a valid rune, we can simply pass it though without further examination, since it cannot be one of the special-chars
we need to actually handle.

This makes it possible to decode mysql outfile data that contains
arbitrary binary data.

Note however, if that that decoding is part of a conversion to CSVs for
later IMPORT, importing that data may still be difficult due to IMPORT’s
handling of non-unicode bytes. A followup could extend the caller, that
uses this decoding to produce CSVs, to additionally re-encode specific
columns to hex to make the end-to-end process including IMPORT smoother.

Release note: none.

dt committed:
cmd/mysqlout: allow hex encoding specified cols

This allows specifying columns which should be encoded as hex as part of
the process of converting from mysql outfile format to CSV.

Our CSV reading, as used in IMPORT, doesn’t handle arbitrary binary very
well. We suggest using hex encoding to get around this, so ideally if a
tool like this is creating CSVs for IMPORT, it should allow using our
recommended encoding.

Release note: none.

mjibson commented:
I don't think we need to merge this. Seems fine to keep it open until full IMPORT support works.

dt commented:
#25615
